be/src/storage/lake/transactions.cpp

```diff
@@ -24,12 +24,56 @@
 
 namespace starrocks::lake {
 
+// check whether all txn log exists, for versions may be published repeatedly and
+// txn log has been deleted when converting from single publish to batch,
+// there we find the latest txnLog as base_version to skip repeated version.
+// result_base_index indicates the index of first txn in the txn_ids, which the corresponding txn_log is not deleted.
+
+// for example:
+// the mode of publish is single,
+// txn2 has been published successfully and visible version in FE is updated to 2,
+// then txn3 is published successfully in BE and the txn_log of txn3 has been deleted, but FE do not get the response for some reason,
+// turn the mode of publish to batch,
+// txn3 ,txn4, txn5 will be published in one publish batch task, so txn3 should be skipped and should return 1, just apply txn_log of txn4 and txn5.
+Status get_base_tablet_metadata_index(TabletManager* tablet_mgr, int64_t tablet_id, int64_t base_version,
+                                      int64_t new_version, std::span<const int64_t>& txn_ids, int& result_base_index) {
+    result_base_index = -1;
+    for (int i = 0; i < txn_ids.size(); i++) {
+        auto txn_id = txn_ids[i];
+        auto log_path = tablet_mgr->txn_log_location(tablet_id, txn_id);
+        auto txn_log_st = tablet_mgr->get_txn_log(log_path, false);
+
+        if (txn_log_st.status().is_not_found()) {
+            auto missig_txn_log_meta = tablet_mgr->get_tablet_metadata(tablet_id, base_version + i + 1);
+            if (missig_txn_log_meta.status().is_not_found()) {
+                // this should't happen
+                LOG(WARNING) << "txn_log of txn: " << txn_id << " not found, and can not find the tablet_meta";
+                return Status::InternalError("Both txn_log and corresponding tablet_meta missing");
+            }
+        } else {
+            result_base_index = i;
+            break;
```

Comment:
It seems to assume if txn log `i` exists, then txn log `i+1`, `i+2`, ... must exist. But this assumption may not hold when cache is present. It's possible txn log `i` and succeeding logs are deleted from object storage, but txn log `i` still resides in local cache and `txn_log_st` is `OK`.

Reply:
The logic there is to deal with the situation `single` switch to `batch`, hence only the first txn_log in txn_ids may be deleted,  so the other txn_logs can not be deleted for they haven't been published?